### ISSUE TITLE
Configuring which TypeScript version is used?

### ISSUE THREAD

**Issue by ffMathy** [codesandbox/codesandbox-client#942]
Can we make it so that I can configure which version of TypeScript to use? Can I (for instance) install `typescript@next` and use that?

**Comment by bengry** [comment]
Is this deployed yet? I can't seem to get CodeSandbox to use any TypeScript version other then `3.3.3`

**Comment by aleclarson** [comment]
@bengry From #1843:
> This doesn't change which version is used by the editor yet, I'm meaning to change this later by introducing a BrowserFS backend that can handle UNPKG.

**Comment by bengry** [comment]
@aleclarson so what does this change?

**Comment by TPReal** [comment]
I guess the issue is that the new syntax cannot be used on the online editor, so this change did not really fix the issue, just prepared for the fix if I understand correctly.

**Comment by ffMathy** [comment]
Shouldn't this be re-opened again then?

**Comment by gaui** [comment]
Could this be re-opened? :pray:

**Comment by ackvf** [comment]
What's the current status? I see that changing `typescript` in `devDependencies` doesn't work?

**Comment by CompuIves** [comment]
It only changes the TypeScript version used by the bundler right now. It's not in the editor yet, as it would require us to dynamically update the VSCode Extensions. I can tackle that one this weekend, it slipped my todo.

**Comment by visionpoint** [comment]
@CompuIves voicing my support for this as well and checking to see if you have any updates on this issue? Cheers

**Comment by mikeplus64** [comment]
> @CompuIves voicing my support for this as well and checking to see if you have any updates on this issue? Cheers

Bump: same =). I have a library that depends on TS 4.0+ and would like to demo it but it's pretty awkward having lots of type errors there.

**Comment by MichaelDeBoey** [comment]
I think this one can be closed since https://github.com/codesandbox/codesandbox-client/pull/4583 is merged.
Can anybody check this out?
I’m currently not on my laptop

**Comment by phryneas** [comment]
@MichaelDeBoey if I interpret this right, adding
```
{
  "dependencies": {
    "typescript": "4.1.2"
  },
}
```

to the `package.json` **should** make the editor use TS 4.1, correct?

Because right now, it seemingly doesn't :disappointed:

**Comment by CompuIves** [comment]
Oh that's right, we don't dynamically set it based on dependencies, but rather based on a string in our code. That's what the mentioned PR fixed. I took a look at the time to see if we could initialize it based on the sandbox config, but I wasn't able to make it work yet as the extension host starts before we fetch the sandbox.

That's the most challenging part, it surely is possible, but we need to restart the extension host whenever the TypeScript version changes (eg. you add it as dependency) or when you change sandbox. Because of this we make the extension host dependent on the sandbox, which will slow the editor initialization for a bit (first we need to fetch the sandbox before we can load the editor). It's something that I think should happen, but it's not implemented yet.

**Comment by tianzhich** [comment]
> Oh that's right, we don't dynamically set it based on dependencies, but rather based on a string in our code. That's what the mentioned PR fixed. I took a look at the time to see if we could initialize it based on the sandbox config, but I wasn't able to make it work yet as the extension host starts before we fetch the sandbox.
> 
> That's the most challenging part, it surely is possible, but we need to restart the extension host whenever the TypeScript version changes (eg. you add it as dependency) or when you change sandbox. Because of this we make the extension host dependent on the sandbox, which will slow the editor initialization for a bit (first we need to fetch the sandbox before we can load the editor). It's something that I think should happen, but it's not implemented yet.

Thank you for your clarify. The [stackblitz](https://github.com/stackblitz/core/issues/483) also has this problem. We just confusing and thought it could be "dynamic" updated cause this issue closed and a fixed pr linked to this...

**Comment by phryneas** [comment]
@CompuIves in the long run, it would be great if we had this option of switching, especially for new TS betas etc. 
In the meantime: could you please bump the TS version up to TypeScript 4.1? That would be a very big short-time win :)

**Comment by CompuIves** [comment]
Yes! PR is here: https://github.com/codesandbox/codesandbox-client/pull/5257

**Comment by OnkelTem** [comment]
I cannot change the version from 4.1.2 to anything else, despite I have 4.2.4 defined in my package.json.
When I click 4.1.2 label in the bottom bar, I see a dialog where I can select only one option: `Use VS Code's version: 4.1.2`.

So how to change the version?

**Comment by JamesACS** [comment]
Duplicate of #5598

**Comment by JamesACS** [comment]
@phryneas Apologies if I'm misunderstanding - it should be possible to add a new Typescript version through the dependencies on a sandbox by sandbox basis

**Comment by barrymay** [comment]
@JamesACS if you're saying that's a feature in place today or there's workarounds for this, it's not true from my experience.

The problem is that the codesandbox doesn't let you choose which TS to you, which leads to those false positives.

**Comment by JamesACS** [comment]
Ah @barrymay I'm with you now - I'll raise this as a feature request

**Comment by IanVS** [comment]
Hi there, it's a shame that https://github.com/stackblitz/core/issues/337 had to be locked due to poor behavior from a user.  But the last status update on that thread was:

>I’m happy to share that there’s a big upgrade coming out later this year that will enable this.

I'm just wondering if there's

1) a way to find the TypeScript version being used
2) a way to change that version

It's a bit confusing to see typescript errors in a project that specifies a version of typescript that supports the syntax I'm using, so I think it would help avoid a lot of confusion if it was clear what version of TS is being used, at least.

**Comment by kc0tlh** [comment]
**Update:** This issue has been resolved in [stackblitz/core/issues/337](https://github.com/stackblitz/core/issues/337#issuecomment-1690364202).

**TL;DR**
When you open a TypeScript file in the [latest StackBlitz editor](https://stackblitz.com/codeflow), the Typescript version is shown in bottom right corner:

![image](https://github.com/codesandbox/codesandbox-client/assets/1532264/0d933731-b439-42e6-b1b9-6a719a7be67f)

You can also  pick the workspace version, allowing you to configure your typescript versions on a per repo basis if desired. Addition context [in the issue](https://github.com/stackblitz/core/issues/337#issuecomment-1690364202) if anyone's interested!

**Comment by CompuIves** [comment]
I think I lost track of this issue, but the new editor of CodeSandbox also supports this. It will use the version of Typescript that you have installed in your workspace. The new editor is enabled by default for Cloud Sandboxes, and we're slowly enabling this as well for browser sandboxes.

**Comment by CompuIves** [comment]
I'll close this issue as soon as we've made the full move, but for now if you want to use it already, we have an option in the editor to convert your sandbox to a cloud sandbox. This would also allow you to get our AI & VSCode integration for that sandbox.

**Comment by LoganDark** [comment]
> The new editor is enabled by default for Cloud Sandboxes, and we're slowly enabling this as well for browser sandboxes.

@CompuIves how do I enable it *now*?

**Comment by CompuIves** [comment]
It should be enabled by default by now! We're also going to release Web VS Code mode soon for all users which allows you to install extensions. If you want to use it already you can do enable it from your experiments (is in the user settings in the editor).

**Comment by LoganDark** [comment]
> It should be enabled by default by now! We're also going to release Web VS Code mode soon for all users which allows you to install extensions. If you want to use it already you can do enable it from your experiments (is in the user settings in the editor).

Unfortunately even with VS Code Web enabled it doesn't seem to have a version selector.

I can go to https://codesandbox.io/p/sandbox/muddy-hill-sjfzwl, focus a TSX file, and whether I have VS Code Web enabled or not, this is all that displays:

![image](https://github.com/codesandbox/codesandbox-client/assets/4723091/2b84ccfe-1e53-4e0f-85f2-586ce3e04379)

and clicking it brings this up:

![image](https://github.com/codesandbox/codesandbox-client/assets/4723091/bbce76d1-7c71-478d-9438-cc91009b63c3)

no way to select the version.

If you check my `tsconfig.json`, I have a compiler plugin that's not being picked up by the language server.

**Comment by CompuIves** [comment]
Interesting, we should use the TypeScript version that's defined in `devDependencies` in package.json for sandboxes. Could you try to change the version of `typescript` in `devDependencies`, refresh, and see if it works?

**Comment by LoganDark** [comment]
@CompuIves How do I verify which version is being used? If you're just cloning the specified version it won't work with compiler plugins, which are what I'm trying to do here (it offers autocomplete for CSS Module imports).

**Comment by phryneas** [comment]
You can use this:
```ts
import { TSVersion } from "@phryneas/ts-version";

type Version = `${TSVersion.Major}.${TSVersion.Minor}`;
```
and then hover `Version`.

**Comment by LoganDark** [comment]
![image](https://github.com/codesandbox/codesandbox-client/assets/4723091/7fade268-43ea-48d0-8f19-0f698dd9edac)
Good to know that it's using a version of TS almost two years out of date and completely ignoring devDependencies which says 5.4.5.

Also, devDependencies are not being installed at all. If I place `@phryneas/ts-version` in `devDependencies` there's no way to install it, even if I manually trigger "install missing dependencies" from the command palette, I just get a toast saying there aren't any missing dependencies. But the moment I move the dependency to regular `dependencies` and save, then it appears instantly without me having to trigger anything. So I tried putting `"typescript": "5.4.5"` in regular `dependencies` also and that does install it but the editor still doesn't use it, it still says 4.9. I even tried refreshing and that didn't help.
